pallets/proxy-teleport/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+#![cfg_attr(not(feature = "std"), no_std)]
+
+type BaseXcm<T> = pallet_xcm::Pallet<T>;
+use frame_support::{
+	dispatch::DispatchResult,
+	ensure,
+	traits::{Contains, EnsureOrigin, Get},
+	weights::Weight,
+};
+use frame_system::pallet_prelude::OriginFor;
+pub use pallet::*;
+use parity_scale_codec::Encode;
+use sp_std::{boxed::Box, vec};
+use xcm::v3::WeightLimit::Unlimited;
+pub use xcm::{
+	opaque::latest::prelude::{Junction, Junctions, MultiLocation, OriginKind},
+	v3::{
+		AssetId, ExecuteXcm, Fungibility,
+		Instruction::{
+			BuyExecution, DepositAsset, DepositReserveAsset, InitiateReserveWithdraw,
+			ReceiveTeleportedAsset, Transact, WithdrawAsset,
+		},
+		MultiAsset, MultiAssetFilter, MultiAssets, Parent, SendXcm, WeightLimit, WildMultiAsset,
+		Xcm, XcmHash,
+	},
+	VersionedMultiAssets, VersionedMultiLocation, VersionedResponse, VersionedXcm,
+};
+
+// #[cfg(test)]
+// mod mock;
+
+// #[cfg(test)]
+// mod tests;
+
+// #[cfg(feature = "runtime-benchmarks")]
+// mod benchmarking;
+// pub mod weights;
+// pub use weights::*;
+
+#[frame_support::pallet]
+pub mod pallet {
+	use super::*;
+	use frame_support::pallet_prelude::*;
+	use frame_system::pallet_prelude::*;
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config + pallet_xcm::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+	}
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// An error ocured during send
+		SendError,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// Execution of an XCM message was attempted.
+		Attempted { outcome: xcm::latest::Outcome },
+		/// A XCM message was sent.
+		Sent {
+			origin: MultiLocation,
+			destination: MultiLocation,
+			message: Xcm<()>,
+			message_id: XcmHash,
+		},
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		#[pallet::call_index(0)]
+		#[pallet::weight(10_000)]
+		pub fn proxy_native_teleport(
+			origin: OriginFor<T>,
+			dest: Box<VersionedMultiLocation>,
+			beneficiary: Box<VersionedMultiLocation>,
+			native_asset_amount: u128,
+			proxy_asset: Box<VersionedMultiAssets>,
+			fee_asset_item: u32,
+		) -> DispatchResult {
+			Self::do_proxy_teleport_assets(
+				origin,
+				dest,
+				beneficiary,
+				native_asset_amount,
+				proxy_asset,
+				fee_asset_item,
+			)
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	fn do_proxy_teleport_assets(
+		origin: OriginFor<T>,
+		dest: Box<VersionedMultiLocation>,
+		beneficiary: Box<VersionedMultiLocation>,
+		native_asset_amount: u128,
+		proxy_asset: Box<VersionedMultiAssets>,
```

Comment:
is this input parameter actually required?

My understanding is that it is always going to be the Relay chain/system  parachain native asset hence the user would always set the same input.

Reply:
good point, I agree it makes sense to leave this flexibility for the future. Only thing, it would be good to document this, particularly that the sovereign account in AH will need to have enough balance of the relevant asset for this to work (apart from it having a liquidity pool vs the native asset etc)